File: LeetCode/Java/69-binary-search-x-的平方根.java
/*
 * @lc app=leetcode.cn id=69 lang=java
 *
 * [69] x 的平方根
 */

// @lc code=start
class Solution {
    public int mySqrt(int x) {
        if (x < 2)
            return x;

        int l = 2;
        int r = x / 2;

        while (l <= r) {
            int m = l + (r - l) / 2;
            long t = (long) m * m;
            if (t > x)
                r = m - 1;
            else if (t < x)
                l = m + 1;
            else
                return m;
        }

        return r;
    }
}
// @lc code=end

File: LeetCode/Python/69-binary-search-x-的平方根.py
#
# @lc app=leetcode.cn id=69 lang=python3
#
# [69] x 的平方根
#


# @lc code=start
class Solution:
    def mySqrt(self, x: int) -> int:
        l, r = 0, x
        ans = -1
        while l <= r:
            m = (l + r) >> 1
            if m * m <= x:
                ans = m
                l = m + 1
            else:
                r = m - 1
        return ans


# @lc code=end
